commit c06e341b3e611f0da2bccec0f8b7f3ad56f1bfde (refs/changes/35/5435/1)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2019-01-25T20:10:26-08:00 (9 months ago)
    
    TRITON-1162 cn-agent dummy backend should have only one way to get sysinfo

diff --git a/lib/backends/dummy/index.js b/lib/backends/dummy/index.js
index c9111bb..7f6f5e3 100644
--- a/lib/backends/dummy/index.js
+++ b/lib/backends/dummy/index.js
@@ -85,6 +85,7 @@ function DummyBackend(opts) {
     var self = this;
 
     assert.object(opts.log, 'opts.log');
+    assert.optionalUuid(opts.serverUuid, 'opts.serverUuid');
 
     self.log = opts.log;
     self.name = opts.backendName;
@@ -111,12 +112,17 @@ DummyBackend.prototype.getSdcConfig = function getSdcConfig(_, callback) {
 };
 
 DummyBackend.prototype.getSysinfo = function getSysinfo(opts, callback) {
+    assert.object(opts, 'opts');
+    assert.object(opts.serverAddress, 'opts.serverAddress');
+    assert.optionalUuid(opts.serverUuid, 'opts.serverUuid');
+
     var self = this;
     var serverUuid = opts.serverUuid || self.serverUuid;
 
     var getter = new sysinfoGetter();
 
     getter.get({
+        serverAddress: opts.serverAddress,
         serverUuid: serverUuid
     }, function _onSysinfo(err, sysinfo) {
         callback(err, sysinfo);
@@ -139,7 +145,10 @@ DummyBackend.prototype.getFirstAdminIp = function getFirstAdminIp(opts, sysinfo,
 
     if (!callback) {
         callback = sysinfo;
-        this.getSysinfo(opts, function (err, sysinfoObj) {
+        this.getSysinfo({
+            serverAddress: {}, // don't need serverAddress, just want Admin IP
+            serverUuid: opts.serverUuid // might be undefined, that's ok
+        }, function (err, sysinfoObj) {
             if (err) {
                 callback(err);
                 return;
@@ -269,6 +278,7 @@ DummyBackend.prototype.getBootTime = function getBootTime(opts, callback) {
     var serverUuid = opts.serverUuid || self.serverUuid;
 
     getter.get({
+        serverAddress: {}, // Don't need server addr to just get Boot Time
         serverUuid: serverUuid
     }, function _onSysinfo(err, sysinfo) {
         var epochTime = jsprim.parseInteger(sysinfo['Boot Time'], {});
diff --git a/lib/backends/dummy/lib/sysinfo.js b/lib/backends/dummy/lib/sysinfo.js
index e2c575c..8bcf1bc 100644
--- a/lib/backends/dummy/lib/sysinfo.js
+++ b/lib/backends/dummy/lib/sysinfo.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2018, Joyent, Inc.
+ * Copyright (c) 2019, Joyent, Inc.
  */
 
 var fs = require('fs');
@@ -20,7 +20,7 @@ function sysinfo() {
 sysinfo.prototype.get = function get(opts, callback) {
     assert.object(opts, 'opts');
     assert.uuid(opts.serverUuid, 'opts.serverUuid');
-    assert.optionalObject(opts.serverAddress, 'opts.serverAddress');
+    assert.object(opts.serverAddress, 'opts.serverAddress');
     assert.func(callback, 'callback');
 
     var filename = common.SERVER_ROOT + '/' + opts.serverUuid + '/sysinfo.json';
@@ -34,11 +34,8 @@ sysinfo.prototype.get = function get(opts, callback) {
         }
 
         sinfo = JSON.parse(data.toString());
-
-        if (opts.serverAddress) {
-            sinfo['CN Agent Port'] = opts.serverAddress.port;
-            sinfo['CN Agent IP'] = opts.serverAddress.address;
-        }
+        sinfo['CN Agent Port'] = opts.serverAddress.port;
+        sinfo['CN Agent IP'] = opts.serverAddress.address;
 
         callback(null, sinfo);
     });
@@ -53,6 +50,7 @@ if (require.main === module) {
     assert.uuid(uuid, 'uuid');
 
     sysinfoGetter.get({
+        serverAddress: {},
         serverUuid: uuid
     }, function onSysinfo(err, info) {
         assert.ifError(err, 'unexpected error loading sysinfo');
diff --git a/lib/backends/dummy/tools/run-servers.js b/lib/backends/dummy/tools/run-servers.js
index 6371086..c91e447 100644
--- a/lib/backends/dummy/tools/run-servers.js
+++ b/lib/backends/dummy/tools/run-servers.js
@@ -34,16 +34,19 @@ function createLog(ctx, callback) {
 }
 
 function loadSysinfo(ctx, callback) {
-    var filename = common.SERVER_ROOT + '/' + ctx.uuid + '/sysinfo.json';
-
-    fs.readFile(filename, function onData(err, data) {
-        if (err) {
-            callback(err);
-            return;
-        }
-
-        ctx.sysinfo = JSON.parse(data.toString());
-        callback();
+    assert.object(ctx, 'ctx');
+    assert.object(ctx.agentserver, 'ctx.agentserver');
+    assert.object(ctx.agentserver.server, 'ctx.agentserver.server');
+    assert.func(ctx.agentserver.server.address,
+        'ctx.agentserver.server.address');
+    assert.func(callback, 'callback');
+
+    ctx.backend.getSysinfo({
+        serverAddress: ctx.agentserver.server.address(),
+        serverUuid: ctx.uuid
+    }, function _onSysinfo(err, sysinfo) {
+        ctx.sysinfo = sysinfo;
+        callback(err);
     });
 }
 
@@ -72,7 +75,7 @@ function createAgentServer(ctx, callback) {
         bindip: ctx.bindIP,
         log: ctx.log,
         port: 0,
-        uuid: ctx.sysinfo.UUID
+        uuid: ctx.uuid
     });
 
     ctx.agentserver.start(function _onStart() {
@@ -140,9 +143,9 @@ function runServer(opts, callback) {
     vasync.pipeline({
         arg: ctx,
         funcs: [
-            loadSysinfo,
             setTaskParams,
-            createAgentServer
+            createAgentServer,
+            loadSysinfo
         ]
     }, function pipelineComplete(err) {
         var app;
diff --git a/package.json b/package.json
index e3224f9..6d1c263 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cn-agent",
   "description": "Triton Compute Node Agent",
-  "version": "2.7.1",
+  "version": "2.7.2",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
